### DISCUSSION TITLE
Create "testsuite" element in test report file per @TestFactory

### DISCUSSION THREAD

**Discussion by rasmussl** [junit-team/junit-framework#3904] [Q&A]
Hi.

I'm looking for a way to have multiple `@TestFactory`'s inside one file (Kotlin) that would resolve in a test report file with a `testsuite` element per TestFactory.
So something like this:
```
<?xml version="1.0" encoding="UTF-8"?>
<testsuites time="6">
    <testsuite name="Tests.One" time="3">
        <testcase name="testCase1" classname="Tests.Registration" time="1" />
        <testcase name="testCase2" classname="Tests.Registration" time="1" />
        <testcase name="testCase3" classname="Tests.Registration" time="1" />
    </testsuite>
    <testsuite name="Tests.Two" time="3">
        <testcase name="testCase4" classname="Tests.Authentication" time="1" />
        <testcase name="testCase5" classname="Tests.Authentication" time="1" />
        <testcase name="testCase6" classname="Tests.Authentication" time="1">
            <failure message="Assertion error message" type="AssertionError">
                <!-- Call stack printed here -->
            </failure>            
        </testcase>
    </testsuite>
</testsuites>
```
as a test report as a result of running this test file:

```
class Tests {

    @TestFactory
    fun `One`(): Stream<DynamicTest> {
         // Some test
    }
    @TestFactory
    fun `Two`(): Stream<DynamicTest> {
         // Some test
    }
}
```
Is it at all possible? If so, then how do you do it?

**Comment by portyu9** [comment]
Hi @rasmussl 

There is no built-in JUnit Jupiter option that maps each `@TestFactory` to a separate legacy XML `<testsuite>`.

A `@TestFactory` is a container in the Jupiter test tree, but JUnit Platform's `LegacyXmlReportGeneratingListener` creates reports at the **`TestPlan` root** level, not at each factory/container level. So two factory methods like `One` and `Two` cannot be configured to become two `<testsuite>` elements in that reporter.

If your reporting consumer supports it, the better current option is **Open Test Reporting**, because it preserves JUnit's hierarchical test structure:

```properties
junit.platform.reporting.open.xml.enabled=true
junit.platform.reporting.output.dir=reports
```

If you specifically need the legacy XML structure shown in the question, you would need to generate/customize the report yourself, for example with a `TestExecutionListener`, grouping each factory's dynamic descendants into the desired `<testsuite>`.

Also note that if the XML file is actually being generated by Gradle, Maven Surefire, or another build tool rather than JUnit Platform's reporting listener, the exact XML grouping is controlled by that tool. In that case the customization has to happen there rather than through `@TestFactory`.

Please mark this answer as accepted if it helped, thank you.
